fix vochaintest races

altergui opened the pull request:
vochaintest is currently plagued by timing issues, where the code is doing time.Sleep() with the hope that a block is mined with a particular transaction included, which frequently fails to happen.

this fixes #575 where i collected a couple of frequent errors, here's my stab at solving all of them

i did refactor major parts of the code, since there were lots of copy&paste and things i have the feeling are not pretty idiomatic like functions embedded inside testVocli for reasons unclear. 

I'd love to refactor more to make the file more readable/maintainable but for now i'll address the urgent issue of the frequent failures

p4u commented on cmd/vochaintest/vochaintest.go:
```diff
@@ -192,6 +198,27 @@ func privKeyToSigner(key string) (*ethereum.SignKeys, error) {
+// If it fails, it will retry 10 times (with 1s interval)
+// before giving up
+func newTestClient(addr string) (*testClient, error) {
+	for tries := 10; tries > 0; tries-- {
```
make this a constant or use the already defined "retries"

altergui replied:
ok, new constant

p4u commented on cmd/vochaintest/vochaintest.go:
```diff
@@ -265,16 +283,34 @@ func censusImport(host string, signer *ethereum.SignKeys) {
+
+		// SetTransactionCost can return OK and then the tx be rejected anyway later.
+		// So, ignore errors, we only care about GetTransactionCost in the next loop
+		_ = mainClient.SetTransactionCost(
```
The error should be checked. 

You are correct, the transaction might fail afterwards (when DeliverTx() is executed) but there might be a first error detected by CheckTx() in the moment the TX reaches the mempool.

altergui replied:
right, error checked

p4u commented on cmd/vochaintest/vochaintest.go:
```diff
@@ -265,16 +283,34 @@ func censusImport(host string, signer *ethereum.SignKeys) {
+func (mainClient testClient) ensureTxCostEquals(signer *ethereum.SignKeys, txType models.TxType, cost uint64) error {
+	for i := 0; i < retries; i++ {
+		// if cost equals expected, we're done
+		newTxCost, _ := mainClient.GetTransactionCost(models.TxType_SET_ACCOUNT_INFO)
```
Do not ignore errors please. Even if they look dummy. Cause afterwards debugging a code that hide some errors is pretty inconvenient.

altergui replied:
done. i added this snippet in some cases where an error is not frequent
```
		if err != nil {
			log.Warn(err)
		}
```
and log.Debug if the error is more common. in any case the loop block should continue trying another iteration.

p4u commented on cmd/vochaintest/vochaintest.go:
```diff
@@ -265,16 +283,34 @@ func censusImport(host string, signer *ethereum.SignKeys) {
-	h, err := mainClient.GetCurrentBlock()
-	if err != nil {
-		return fmt.Errorf("cannot get current height: %w", err)
+func (mainClient testClient) ensureTxCostEquals(signer *ethereum.SignKeys, txType models.TxType, cost uint64) error {
```
shouldn't `testClient`be a pointer? @mvdan

altergui replied:
true, good catch, fixed

p4u commented on cmd/vochaintest/vochaintest.go:
```diff
@@ -345,71 +383,83 @@ func initAccounts(treasurer, oracle string, accountKeys []*ethereum.SignKeys, ho
+func (mainClient testClient) ensureAccountInfoEquals(account *ethereum.SignKeys, infoURI string) error {
+	for i := 0; i < retries; i++ {
+		// if account InfoURI is as expected, we're done
+		acct, _ := mainClient.GetAccount(account.Address())
```
error?

p4u commented on cmd/vochaintest/vochaintest.go:
```diff
@@ -345,71 +383,83 @@ func initAccounts(treasurer, oracle string, accountKeys []*ethereum.SignKeys, ho
+		// else, set it
+		// CreateOrSetAccount can return OK and then the tx be rejected anyway later.
+		// So, ignore errors, we only care about GetAccount in the next loop
+		_ = mainClient.CreateOrSetAccount(account, common.Address{}, infoURI, acct.Nonce, nil)
```
error

altergui committed:
client: total refactor (fix #299, #497, #575)

client/api: don't panic when database is locked
client/api: c.TestResults() cleaner code and use c.WaitUntilNextBlock()
client/api: c.TestPreRegisterKeys() use c.WaitUntilProcessAvailable and c.WaitUntilNextBlock
client/api: c.TestSendVotes() use c.WaitUntilProcessReady and c.WaitUntilProcessKeys
client/api: c.TestSendVotes() don't panic when c.Request returns EOF
client/api: c.TestSendVotes() cleaner code: deintend else block
client/api: c.TestSendAnonVotes() use c.WaitUntilProcessReady and c.WaitUntilEnvelopeHeight
client/api: c.CreateProcess() minimum startBlockIncrement >= 5 with EncryptedVotes (fix #299)
client/api: c.CreateProcess() use c.WaitUntilProcessAvailable
client/api: document SetProcessStatus and EndProcess
client/api: c.SetTransactionCost() now returns txHash
client/api: lint c.CreateOrSetAccount(), c.MintTokens(), c.SetOracle()
client/api: c.SendTokens(), c.SetAccountDelegate(), c.CollectFaucet() now return txHash
client/api: new func c.GetTxByHash()
client/util: new funcs WaitUntilNBlocks, WaitUntilNextBlock, WaitUntilTxMined, WaitUntilProcessAvailable, WaitUntilProcessReady, WaitUntilEnvelopeHeight, WaitUntilProcessKeys

altergui committed:
testsuite: never run vocli and tokentransactions concurrently

altergui committed:
vochaintest: total refactor (fix #299, #497, #575)

 * introduce testClient struct: wraps client.Client and attaches ensure* methods to it
 * ensureTxCostEquals() robustly replaces mainClient.SetTransactionCost()
 * ensureProcessCreated() uses WaitUntilNextBlock and WaitUntilProcessAvailable
 * ensureProcessEnded() uses WaitUntilNextBlock
 * ensureProcessEnded() fixed parentheses which lead to nil pointer dereference
 * new func ensureAccountInfoEquals()
 * mkTreeAnonVoteTest() use WaitUntilProcessAvailable
 * testSetTxCost() use ensureTxCostEquals
 * testCreateAndSetAccount() use ensureAccountExists, ensureAccountHasTokens, ensureAccountInfoEquals
 * testSendTokens() use WaitUntilTxMined
 * testSetAccountDelegate() use WaitUntilTxMined
 * testCollectFaucet() use WaitUntilTxMined
 * testVocli: new funcs vocliEnsureAccountHasBalance, vocliRepeatCmdUntilStdoutContains, vocliWaitUntilAccountInfoContains, vocliWaitUntilAccountExists
 * testVocli: replace many time.Sleep(time.Second * 10) with those robust alternatives
 * testVocli: move executeCommand out of testVocli
